toad-common/src/array.rs

```diff
@@ -0,0 +1,229 @@
+  ///
+  /// The default implementation invokes `Default::default`
+  fn reserve(_: usize) -> Self {
+    Default::default()
+  }
+}
+
+impl<T> Reserve for Vec<T> {
+  fn reserve(n: usize) -> Self {
+    Self::with_capacity(n)
+  }
+}
+
+impl<A: tinyvec::Array> Reserve for tinyvec::ArrayVec<A> {}
+
+/// An ordered indexable collection of some type `Item`
+///
+/// # Provided implementations
+/// - [`Vec`]
+/// - [`tinyvec::ArrayVec`]
+///
+/// Notably, not `heapless::ArrayVec` or `arrayvec::ArrayVec`. An important usecase within `toad`
+/// is [`Extend`]ing the collection, and the performance of `heapless` and `arrayvec`'s Extend implementations
+/// are notably worse than `tinyvec`.
+///
+/// `tinyvec` also has the added bonus of being 100% unsafe-code-free, meaning if you choose `tinyvec` you eliminate the
+/// possibility of memory defects and UB.
+///
+/// # Requirements
+/// - [`Default`] for creating the collection
+/// - [`Extend`] for mutating and adding onto the collection (1 or more elements)
+/// - [`Reserve`] for reserving space ahead of time
+/// - [`GetSize`] for bound checks, empty checks, and accessing the length
+/// - [`FromIterator`] for [`collect`](core::iter::Iterator#method.collect)ing into the collection
+/// - [`IntoIterator`] for iterating and destroying the collection
+/// - [`Deref<Target = [T]>`](Deref) and [`DerefMut`] for:
+///    - indexing ([`Index`](core::ops::Index), [`IndexMut`](core::ops::IndexMut))
+///    - iterating ([`&[T].iter()`](primitive@slice#method.iter) and [`&mut [T].iter_mut()`](primitive@slice#method.iter_mut))
+pub trait Array:
+  Default
+  + GetSize
+  + Reserve
+  + Deref<Target = [<Self as Array>::Item]>
+  + DerefMut
+  + Extend<<Self as Array>::Item>
+  + FromIterator<<Self as Array>::Item>
+  + IntoIterator<Item = <Self as Array>::Item>
+{
+  /// The type of item contained in the collection
+  type Item;
+
+  /// Insert a value at a particular index of a collection.
+  fn insert_at(&mut self, index: usize, value: <Self as Array>::Item);
+
+  /// Try to remove an entry from the collection.
+  ///
+  /// Returns `Some(Self::Item)` if `index` was in-bounds, `None` if `index` is out of bounds.
+  fn remove(&mut self, index: usize) -> Option<<Self as Array>::Item>;
+
+  /// Add a value to the end of a collection.
+  fn push(&mut self, value: <Self as Array>::Item);
+}
+
+/// Collections that support extending themselves mutably from copyable slices
+pub trait AppendCopy<T: Copy> {
+  /// Extend self mutably, copying from a slice.
+  ///
+  /// Worst-case implementations copy 1 element at a time (time O(n))
+  ///
+  /// Best-case implementations copy as much of the origin slice
+  /// at once as possible (system word size), e.g. [`Vec::append`].
+  /// (still linear time, but on 64-bit systems this is 64 times faster than a 1-by-1 copy.)
+  fn append_copy<'a>(&mut self, i: &[T]);
+}
+
+impl<T: Copy> AppendCopy<T> for Vec<T> {
+  fn append_copy<'a>(&mut self, i: &[T]) {
+    self.extend(i);
```

Comment:
pinning the type here to a slice allows Vec's specialized extend implementation to copy the slice to the end of the vec with frightening efficiency